### PR TITLE
[ZenTorch] Testing and CI

### PR DESCRIPTION
**Pull request by IlyasMoutawwakil**
This PR adds tests and CI for the custom `torch.compile` backend `zentorch`.
Tests are using tiny models and are mostly checking computational sanity between `inductor` and `zentorch` backends.
The workflow is using the `amdrun` cli command which was modified to set optimal `zentorch` env vars as well as `rocm`.

### CODE REVIEW

**Comment by IlyasMoutawwakil** [comment]
I will add some details to the readme/docs and a dockerfile once zendnn plugin is installable.

File: tests/zentorch/test_sanity.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    assert os.environ["OMP_DYNAMIC"] == "False"
+    assert os.environ["ZENDNN_GEMM_ALGO"] == "4"
+    assert os.environ["OMP_WAIT_POLICY"] == "ACTIVE"
+    assert os.environ["OMP_NUM_THREADS"] == f"{os.cpu_count()}"
+    assert os.environ["GOMP_CPU_AFFINITY"] == f"0-{os.cpu_count() - 1}"
+    assert os.environ["LD_PRELOAD"] == "/usr/lib/x86_64-linux-gnu/libjemalloc.so"
+    assert (
+        os.environ["MALLOC_CONF"]
+        == "oversize_threshold:1,background_thread:true,metadata_thp:auto,dirty_decay_ms:-1,muzzy_decay_ms:-1"
+    )
```
Comment (IlyasMoutawwakil) [inline tests/zentorch/test_sanity.py:31]: tests if `amdrun` sets the right env vars

File: tests/zentorch/testing_utils.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+        "hf-internal-testing/tiny-random-Blip2ForConditionalGeneration": ["image-to-text", "visual-question-answering"]
+    },
+    # automatic speech recognition
+    "whisper": {"openai/whisper-tiny": ["automatic-speech-recognition"]},
+}
+
+SUPPORTED_DIFFUSION_PIPELINES_TINY = {
+    "stable-diffusion": {"hf-internal-testing/tiny-stable-diffusion-torch": ["text-to-image"]},
+    "stable-diffusion-xl": {"hf-internal-testing/tiny-stable-diffusion-xl-pipe": ["text-to-image"]},
+}
```
Comment (IlyasMoutawwakil) [inline tests/zentorch/testing_utils.py:84]: List of tested architectures x models x tasks

File: docker/transformers-pytorch-amd-cpu-zentorch/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    git \
+    ffmpeg \
+    python3.8 \
+    python3-pip \
+    python3.8-dev \
+    build-essential \
+    libjemalloc-dev && \
+    apt-get clean && \
+    rm -rf /var/lib/apt/lists/* && \
+    update-alternatives --install /usr/bin/python python /usr/bin/python3.8 1 && \
```
Comment (fxmarty) [inline docker/transformers-pytorch-amd-cpu-zentorch/Dockerfile:34]: Maybe just to keep in mind that python 3.8 soon reaches end-of-life

File: optimum/amd/cli.py [hunk 4]
```diff
@@ -1,47 +1,111 @@
 import os
+import subprocess
 import sys
-from argparse import ArgumentParser
 
-from .topology_utils import extract_max_avg_bandwidth_cluster, get_bandwidth_matrix
 
+try:
```
Comment (fxmarty) [inline optimum/amd/cli.py:6]: All of this could be worth moving to Accelerate no? They have their own `accelerate launch` which is more widely used & advertised in Transformers doc/examples.

File: tests/zentorch/test_sanity.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+
+            model = load_transformers_model(model_id, task)
+            inductor_model = compile_transformers_model(model, backend="inductor")
+            inductor_logits = inductor_model.forward(**inputs).logits
+
+            model = load_transformers_model(model_id, task)
+            zentorch_model = compile_transformers_model(model, backend="zentorch")
+            zentorch_logits = zentorch_model.forward(**inputs).logits
+
+            torch.testing.assert_close(inductor_logits, zentorch_logits, rtol=1e-3, atol=1e-5)
```
Comment (fxmarty) [inline tests/zentorch/test_sanity.py:50]: I don't understand the motivation to compare inductor vs zentorch. Why not PyTorch eager vs zentorch?

File: tests/zentorch/testing_utils.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    return model
+
+
+def compile_diffusion_pipeline(pipeline, backend):
+    torch._dynamo.reset()
+
+    pipeline.unet = torch.compile(pipeline.unet, backend=backend)
+    pipeline.vae.decoder = torch.compile(pipeline.vae.decoder, backend=backend)
+
+    return pipeline
```
Comment (fxmarty) [inline tests/zentorch/testing_utils.py:171]: I think it would be nice to add tests with `fullgraph=True` as well, but that can be done later

File: tests/zentorch/testing_utils.py [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+    elif task == "automatic-speech-recognition":
+        return AutoModelForSpeechSeq2Seq.from_pretrained(model_id)
+    elif task == "audio-classification":
+        return AutoModelForAudioClassification.from_pretrained(model_id)
+    elif task == "image-to-text":
+        return AutoModelForVision2Seq.from_pretrained(model_id)
+    elif task == "visual-question-answering":
+        return AutoModelForVisualQuestionAnswering.from_pretrained(model_id)
+    else:
+        raise ValueError(f"Task {task} not supported")
```
Comment (fxmarty) [inline tests/zentorch/testing_utils.py:105]: Why not use `TasksManager.get_model_from_task`?

File: tests/zentorch/testing_utils.py [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+
+def load_diffusion_pipeline(pipeline_id: str, task: str):
+    if task == "text-to-image":
+        return DiffusionPipeline.from_pretrained(pipeline_id)
+    else:
+        raise ValueError(f"Task {task} not supported")
```
Comment (fxmarty) [inline tests/zentorch/testing_utils.py:112]: same

File: tests/zentorch/test_sanity.py [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+            inductor_model = compile_transformers_model(model, backend="inductor")
+            inductor_logits = inductor_model.generate(**inputs, **FAST_TEXT_GENERATION_KWARGS).logits
+            inductor_logits = torch.stack(inductor_logits, dim=1)
+
+            model = load_transformers_model(model_id, task)
+            zentorch_model = compile_transformers_model(model, backend="zentorch")
+            zentorch_logits = zentorch_model.generate(**inputs, **FAST_TEXT_GENERATION_KWARGS).logits
+            zentorch_logits = torch.stack(zentorch_logits, dim=1)
+
+            torch.testing.assert_close(inductor_logits, zentorch_logits, rtol=1e-3, atol=1e-5)
```
Comment (fxmarty) [inline tests/zentorch/test_sanity.py:71]: Generation tests on tiny model can sometimes pass but not on real size models (from experience with ORT). It would be nice in the future to add nightly tests with real sized models.

File: tests/zentorch/test_sanity.py [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+
+
+@pytest.mark.parametrize("model_type", SUPPORTED_SIMPLE_MODELS_TINY.keys())
+def test_simple_models(model_type: str):
+    model_id_and_tasks = SUPPORTED_SIMPLE_MODELS_TINY[model_type]
+
+    for model_id, tasks in model_id_and_tasks.items():
+        for task in tasks:
+            inputs = get_transformers_model_inputs(model_id, task)
+
```
Comment (fxmarty) [inline tests/zentorch/test_sanity.py:41]: I would add a `model.eval()` & `with torch.no_grad():`

Actually, it could be helpful to have a test as well about gradients, as compile is about but the forward & backward (can be done later of course)

**Review by fxmarty — COMMENTED** [review]
@IlyasMoutawwakil Sorry for the late review I missed the notification

File: docker/transformers-pytorch-amd-cpu-zentorch/Dockerfile (hunk 3 above)
Comment (IlyasMoutawwakil) [reply docker/transformers-pytorch-amd-cpu-zentorch/Dockerfile:34]: the plugin we have right now is compiled with Python3.8
https://huggingface.slack.com/archives/C06GVGQA0JG/p1710936009492349

File: optimum/amd/cli.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply optimum/amd/cli.py:6]: yes it makes sense for distributed inference/training but not sure about zentorch environment variables.

File: setup.py [hunk 11]
```diff
@@ -27,13 +27,16 @@
...
+    "pytest-xdist",
+    "diffusers",
 ]
 
 QUALITY_REQUIRE = ["black~=23.1", "ruff>=0.0.241,<=0.0.259"]
 
 EXTRAS_REQUIRE = {
     "quality": QUALITY_REQUIRE,
     "tests": TESTS_REQUIRE,
+    "zentorch": ["zentorch", "torch==2.2.1"],
```
[the remark this answers is not in the document]
Comment (IlyasMoutawwakil) [reply setup.py:39]: thanks didn't think of that

File: tests/zentorch/test_sanity.py (hunk 5 above)
Comment (IlyasMoutawwakil) [reply tests/zentorch/test_sanity.py:50]: https://huggingface.slack.com/archives/C06GVGQA0JG/p1710338953229219?thread_ts=1709831567.529199&cid=C06GVGQA0JG

File: tests/zentorch/testing_utils.py (hunk 8 above)
Comment (IlyasMoutawwakil) [reply tests/zentorch/testing_utils.py:111]: it was mostly because for diffusion pipelines I wanted to test both pipelines on the same task `text-to-image` but the taks manager has different classification (stable-diffusion is the considered a task and model type) and the returned class for xl variant was `StableDiffusionXLImg2ImgPipeline`
https://github.com/huggingface/optimum/blob/2f75b0da85b95030b82b281205aa4284df2f998a/optimum/exporters/tasks.py#L189-L192

File: tests/zentorch/test_sanity.py (hunk 9 above)
Comment (IlyasMoutawwakil) [reply tests/zentorch/test_sanity.py:71]: yes these are fast/tiny tests, we'll slow ones later

File: tests/zentorch/test_sanity.py (hunk 10 above)
Comment (IlyasMoutawwakil) [reply tests/zentorch/test_sanity.py:41]: just realized I removed the `inference_mode` context at a certain point. will put it back.

**Review by mht-sharma — COMMENTED** [review]
Great work Ilyas, left nit comments

File: tests/zentorch/testing_utils.py (hunk 7 above)
Comment (mht-sharma) [reply tests/zentorch/testing_utils.py:104]: Probably ok to have not have dependency on optimum.exporters unless needed?

File: tests/zentorch/test_sanity.py (hunk 5 above)
Comment (fxmarty) [reply tests/zentorch/test_sanity.py:50]: Okay, not sure why but okay

File: tests/zentorch/testing_utils.py (hunk 7 above)
Comment (fxmarty) [reply tests/zentorch/testing_utils.py:104]: I think it's fine (especially in tests), tasks.py don't require any dependency apart from transformers/huggingface_hub

File: optimum/amd/cli.py [hunk 12]
```diff
@@ -1,47 +1,111 @@
...
-    assert len(sys.argv) >= 4, "Usage: amdrun --ngpus <num_gpus> <script> <script_args>"
+    CPU_COUNT = os.cpu_count()
+    MALLOC_CONF = "oversize_threshold:1,background_thread:true,metadata_thp:auto,dirty_decay_ms:-1,muzzy_decay_ms:-1"
+    LD_PRELOAD = "/usr/lib/x86_64-linux-gnu/libjemalloc.so"
+    GOMP_CPU_AFFINITY = f"0-{CPU_COUNT - 1}"
+    OMP_NUM_THREADS = f"{CPU_COUNT}"
+    OMP_WAIT_POLICY = "ACTIVE"
+    ZENDNN_GEMM_ALGO = "4"
+    OMP_DYNAMIC = "False"
+
```
[the remark this answers is not in the document]
Comment (IlyasMoutawwakil) [reply optimum/amd/cli.py:38]: Will ask EPYC team for some doc/explanation of why we use these values specifically.
The only one that might need to change here is `LD_PRELOAD` because we can't be sure where the `jemalloc` binary is, I can add a check that this file exist and show a warning suggesting setting it manually (like `LD_PRELOAD=... amdrun python ...`. the others were suggested as is by the EPYC team.

**Comment by IlyasMoutawwakil** [comment]
diffusers tests are passing again (it was a seed issue)

File: optimum/amd/cli.py (hunk 4 above)
Comment (fxmarty) [reply optimum/amd/cli.py:9]: why not?

**Review by fxmarty — APPROVED** [review]
LGTM great work!

File: optimum/amd/cli.py (hunk 4 above)
Comment (IlyasMoutawwakil) [reply optimum/amd/cli.py:9]: I thought `launch` can only be used in a distributed setting, but looking at the docs I guess not necessarily.

**Review by mht-sharma — APPROVED** [review]
LGTM! Great work!

**Comment by IlyasMoutawwakil** [comment]
Thanks for the reviews !
